clients/admin-ui/src/features/common/custom-fields/hooks.ts

```diff
@@ -0,0 +1,191 @@
+            })),
+          },
+        ])
+      ),
+    [allAllowListQuery.data]
+  );
+
+  const idToCustomFieldDefinition = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldDefinitionsQuery.data).map((cfd) => [
+          cfd.id,
+          cfd,
+        ])
+      ),
+    [customFieldDefinitionsQuery]
+  );
+
+  const definitionIdToCustomField = useMemo(
+    () =>
+      new Map(
+        filterWithId(customFieldsQuery.data).map((fd) => [
+          fd.custom_field_definition_id,
+          fd,
+        ])
+      ),
+    [customFieldsQuery]
+  );
+
+  const sortedCustomFieldDefinitionIds = useMemo(() => {
+    const ids = [...idToCustomFieldDefinition.keys()];
+    ids.sort();
+    return ids;
+  }, [idToCustomFieldDefinition]);
+
+  /**
+   * Issue a batch of upsert and delete requests that will sync the form selections to the
+   * backend.
+   */
+  const upsertCustomFields = useCallback(
+    async (formValues: CustomFieldsFormValues) => {
+      // When creating an resource, the fides key may have initially been blank. But by the time the
+      // form is submitted it must not be blank (not undefined, not an empty string).
+      const fidesKey = formValues.fides_key || resourceFidesKey;
+      if (!fidesKey) {
+        return;
+      }
+
+      const { definitionIdToCustomFieldValue } = formValues;
+
+      // This will be undefined if the form never rendered a `CustomFieldList` that would assign
+      // form values.
+      if (!definitionIdToCustomFieldValue) {
+        return;
+      }
+
+      try {
+        // This would be a lot simpler (and more efficient) if the API had an endpoint for updating
+        // all the metadata associated with a field, including deleting options that weren't passed.
+        await Promise.allSettled(
+          sortedCustomFieldDefinitionIds.map((definitionId) => {
+            const customField = definitionIdToCustomField.get(definitionId);
+            const value = definitionIdToCustomFieldValue[definitionId];
+
+            if (
+              value === undefined ||
+              value === "" ||
+              (Array.isArray(value) && value.length === 0)
+            ) {
+              if (!customField?.id) {
+                return undefined;
+              }
+
+              return deleteCustomFieldMutationTrigger(customField);
+            }
+
+            const body = {
+              custom_field_definition_id: definitionId,
+              resource_id: fidesKey,
+              id: customField?.id,
+              value,
+            };
+
+            return upsertCustomFieldMutationTrigger(body);
+          })
+        );
+
+        successAlert(
+          `Custom field(s) successfully saved and added to this ${resourceType} form.`
+        );
+      } catch (e) {
+        errorAlert(
+          `One or more custom fields have failed to save, please try again.`
+        );
+        // eslint-disable-next-line no-console
+        console.error(e);
```

Comment:
Can you remove this console error statement?

Reply:
I was thinking it could be helpful even after merge in case we encounter another API incompatibility.